improvement(remoter): re-runing command on retriable exceptions

bentsi opened the pull request:
We can safely retry the command when it didn't run on remote.
This situation can happen when SSH/channel connection was not
successfully initiated.
Related issues: #1793, #1631, #1815

## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I gave variables/functions meaningful self-explanatory names
- [x] I didn't leave commented-out/debugging code
- [x] I didn't copy-paste code
- [x] I added the relevant `backport` labels
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

yarongilor commented on sdcm/remote.py:
```diff
@@ -53,10 +53,28 @@ class SSHConnectTimeoutError(Exception):
+            return True
+    return False
+
+
```
perhaps an idea for improvement, generalize this to all exceptions and errors, could be writing the executed requested command and its timestamp to a file on the remote node after the command was executed. this gives more certainty if the command is already executed or can be retried for many more types of connection errors.

bentsi replied:
As I said in one of my comments, verifying the reliable protocol will not give any advantage. If there is a network issue you need to know about it and resolve the root cause.

ShlomiBalalis commented on sdcm/remote.py:
```diff
@@ -280,9 +299,11 @@ def _run():
+            except NETWORK_EXCEPTIONS as ex:
                 LOGGER.error(ex)
                 self._ssh_is_up.clear()
+                if is_exception_retriable(str(ex)):
```
I assume that on the two other exceptions are different? Because on the other two instances you called is_exception_retriable using ex.result.stderr, while here you just use str(ex)

bentsi replied:
I will rename is_exception_retriable to is_error_retriable. Will that be more clear?
Other two instances run rsync/scp commands using local runner, that has nothing related to SSH. I am just parsing stderr of the rsync/scp commands

ShlomiBalalis replied:
Yep, that'll be better

bentsi replied:
done

bentsi commented:
not backporting this, since it depends on 3 other commits. If we see such issues happening again I will make an effort to backport this.

yarongilor commented:
@bentsi this is reproduced all the time in branch-3.3.
see last run of 1TB:  https://jenkins.scylladb.com/job/scylla-3.3/job/longevity/job/longevity-1tb-7days/8/

roydahan commented:
@bentsi please backport this to 3.3

amoskong commented:
@bentsi 

We need this PR for branch-2019.1

My staging upgrade job of 2019.1 failed for https://github.com/scylladb/scylla-cluster-tests/issues/1815

```
ssh_exchange_identification: Connection closed by remote host
```

amoskong commented:
> @bentsi
> 
> We need this PR for branch-2019.1
> 
> My staging upgrade job of 2019.1 failed for #1815
> 
> ```
> ssh_exchange_identification: Connection closed by remote host
> ```

There are some conflicts in cherry-pick commit 5503f25deee5b08e3216fb8b6343a2291d515ab7 to branch-2019.1

bentsi commented:
in order to backport this need to backport the following chain of commits:
```
* 5503f25d - improvement(remoter): re-runing command on retriable exceptions (3 months ago) <Bentsi Magidovich>
* f55be1ad - (improve-search-builder-test-id) fix(remote.py): disable stdin captureing (#1791) (3 months ago) <Israel Fruchter>
* 7325ee62 - feature(loaders): netflix's ndbench as a new loader tool (3 months ago) <Israel Fruchter>
* 27ea4e1d - improvement(remoter): moved NETWORK_EXCEPTIONS to the remoter, added stop method and improved run (3 months ago) <Bentsi Magidovich>
* 8ca028b7 - Add new_session to remoter (4 months ago) <Dmitry Kropachev>
* ee10e43f - Move is_up into retry cycle (4 months ago) <Dmitry Kropachev>
* 53393fc8 - Fix LocalRunner to debug output when verbose=True (4 months ago) <Israel Fruchter>
* 83d50123 - Remoter: Setting SSH key type as RSA explicitly (4 months ago) <Bentsi Magidovich>
* db6a2fb5 - remoter: ssh_up thread uses it's own connection object to address the issue with paramiko (5 months ago) <Bentsi Magidovich>
* 7d139572 - Remoter: added ability to compeltely silence the remoter (6 months ago) <Bentsi Magidovich>
* 5b931f10 - cluster_aws: retry to stop scylla-server after instance restart (#1469) (6 months ago) <Amos Kong>
* eb463332 - fix(logcollector): Collectlogs and restore monitor stack for large data (4 weeks ago) <Aleksandr Bykov>
* 43240eef - improvement(remoter): add sleep time between SSH authorization attempts (6 weeks ago) <Bentsi Magidovich>
* ec20d478 - refactor(tagging): generalize tagging and refactor Docker backend (7 weeks ago) <Evgeniy Naydanov>
* 660dad5e - refactor(tagging): add ContainerManager (7 weeks ago) <Evgeniy Naydanov>
* 7b189596 - improvement(remoter): making socket time out retriable (8 weeks ago) <Bentsi Magidovich>
* ff9c1a78 - fix(docker): add retry to check_aio_max_nr() (8 weeks ago) <Evgeniy Naydanov>
* 74e570cd - fix(sdcm/docker.py): fix docker and non-docker teardown proceedure (3 months ago) <Dmitry Kropachev>
* 5503f25d - improvement(remoter): re-runing command on retriable exceptions (3 months ago) <Bentsi Magidovich>
* f55be1ad - (improve-search-builder-test-id) fix(remote.py): disable stdin captureing (#1791) (3 months ago) <Israel Fruchter>
* 7325ee62 - feature(loaders): netflix's ndbench as a new loader tool (3 months ago) <Israel Fruchter>
* 27ea4e1d - improvement(remoter): moved NETWORK_EXCEPTIONS to the remoter, added stop method and improved run (3 months ago) <Bentsi Magidovich>
* 8ca028b7 - Add new_session to remoter (4 months ago) <Dmitry Kropachev>
* ee10e43f - Move is_up into retry cycle (4 months ago) <Dmitry Kropachev>
* 53393fc8 - Fix LocalRunner to debug output when verbose=True (4 months ago) <Israel Fruchter>
* 83d50123 - Remoter: Setting SSH key type as RSA explicitly (4 months ago) <Bentsi Magidovich>
* db6a2fb5 - remoter: ssh_up thread uses it's own connection object to address the issue with paramiko (5 months ago) <Bentsi Magidovich>
* 7d139572 - Remoter: added ability to compeltely silence the remoter (6 months ago) <Bentsi Magidovich>
* 5b931f10 - cluster_aws: retry to stop scylla-server after instance restart (#1469) (6 months ago) <Amos Kong>
```
I don't think it is feasible. Meybe we need to check an option to update 2019.1 branch.

bentsi commented:
@roydahan maybe we should update the 2019.1 branch( for instance to 4.0)?

amoskong commented:
> in order to backport this need to backport the following chain of commits:
> 
> ```
> * 5503f25d - improvement(remoter): re-runing command on retriable exceptions (3 months ago) <Bentsi Magidovich>
> * f55be1ad - (improve-search-builder-test-id) fix(remote.py): disable stdin captureing (#1791) (3 months ago) <Israel Fruchter>
> * 7325ee62 - feature(loaders): netflix's ndbench as a new loader tool (3 months ago) <Israel Fruchter>
> * 27ea4e1d - improvement(remoter): moved NETWORK_EXCEPTIONS to the remoter, added stop method and improved run (3 months ago) <Bentsi Magidovich>
> * 8ca028b7 - Add new_session to remoter (4 months ago) <Dmitry Kropachev>
> * ee10e43f - Move is_up into retry cycle (4 months ago) <Dmitry Kropachev>
> * 53393fc8 - Fix LocalRunner to debug output when verbose=True (4 months ago) <Israel Fruchter>
> * 83d50123 - Remoter: Setting SSH key type as RSA explicitly (4 months ago) <Bentsi Magidovich>
> * db6a2fb5 - remoter: ssh_up thread uses it's own connection object to address the issue with paramiko (5 months ago) <Bentsi Magidovich>
> * 7d139572 - Remoter: added ability to compeltely silence the remoter (6 months ago) <Bentsi Magidovich>
> * 5b931f10 - cluster_aws: retry to stop scylla-server after instance restart (#1469) (6 months ago) <Amos Kong>
> * eb463332 - fix(logcollector): Collectlogs and restore monitor stack for large data (4 weeks ago) <Aleksandr Bykov>
> * 43240eef - improvement(remoter): add sleep time between SSH authorization attempts (6 weeks ago) <Bentsi Magidovich>
> * ec20d478 - refactor(tagging): generalize tagging and refactor Docker backend (7 weeks ago) <Evgeniy Naydanov>
> * 660dad5e - refactor(tagging): add ContainerManager (7 weeks ago) <Evgeniy Naydanov>
> * 7b189596 - improvement(remoter): making socket time out retriable (8 weeks ago) <Bentsi Magidovich>
> * ff9c1a78 - fix(docker): add retry to check_aio_max_nr() (8 weeks ago) <Evgeniy Naydanov>
> * 74e570cd - fix(sdcm/docker.py): fix docker and non-docker teardown proceedure (3 months ago) <Dmitry Kropachev>
> * 5503f25d - improvement(remoter): re-runing command on retriable exceptions (3 months ago) <Bentsi Magidovich>
> * f55be1ad - (improve-search-builder-test-id) fix(remote.py): disable stdin captureing (#1791) (3 months ago) <Israel Fruchter>
> * 7325ee62 - feature(loaders): netflix's ndbench as a new loader tool (3 months ago) <Israel Fruchter>
> * 27ea4e1d - improvement(remoter): moved NETWORK_EXCEPTIONS to the remoter, added stop method and improved run (3 months ago) <Bentsi Magidovich>
> * 8ca028b7 - Add new_session to remoter (4 months ago) <Dmitry Kropachev>
> * ee10e43f - Move is_up into retry cycle (4 months ago) <Dmitry Kropachev>
> * 53393fc8 - Fix LocalRunner to debug output when verbose=True (4 months ago) <Israel Fruchter>
> * 83d50123 - Remoter: Setting SSH key type as RSA explicitly (4 months ago) <Bentsi Magidovich>
> * db6a2fb5 - remoter: ssh_up thread uses it's own connection object to address the issue with paramiko (5 months ago) <Bentsi Magidovich>
> * 7d139572 - Remoter: added ability to compeltely silence the remoter (6 months ago) <Bentsi Magidovich>
> * 5b931f10 - cluster_aws: retry to stop scylla-server after instance restart (#1469) (6 months ago) <Amos Kong>
> ```
> 
> I don't think it is feasible. Meybe we need to check an option to update 2019.1 branch.
> @roydahan maybe we should update the 2019.1 branch( for instance to 4.0)?

ping @roydahan , what's your opinion?

----
BTW, what's the status of Release 3.3.2, anything still needs from my side?

The ssh/remoter issue in SCT caused one job failed in the past, but it passed in other job. So I sent the PASSed upgrade result by email.

Just insure that we hadn't fixed the ssh/remoter issue in SCT 3.3 for complex backport conflicts, and we don't need to fix it for release 3.3.
